src/main/java/com/aws/greengrass/deployment/DeploymentQueue.java

```diff
@@ -6,76 +6,162 @@
+ *     - offered.getDeploymentType().equals(SHADOW)</p>
+ *
+ * <p>When an offered deployment is a shadow deployment:
+ *     - if there is already a shadow deployment in the queue, then replace it and preserve queue order.
+ *     - otherwise, enqueue normally.</p>
+ */
 public class DeploymentQueue {
 
+    /**
+     * For the internal queue, shadow deployments use a special queue id. Effectively, this means there can be at most
+     * one shadow deployment in the queue.
+     */
+    private static final String SHADOW_DEPLOYMENT_QUEUE_ID = Deployment.DeploymentType.SHADOW.name();
+
+    private static final String DEPLOYMENT_ID_LOG_KEY = "DeploymentId";
+    private static final String DISCARDED_DEPLOYMENT_ID_LOG_KEY = "DiscardedDeploymentId";
     private static final Logger logger = LogManager.getLogger(DeploymentQueue.class);
-    private final ConcurrentLinkedQueue<Deployment> deploymentsQueue = new ConcurrentLinkedQueue();
 
     /**
-     * Add a deployment to the queue.
+     * Internal queue of deployment id's. For shadow deployments, the internal queue id is not the same as the
+     * actual deployment id.
+     */
+    private final Queue<String> deploymentIdQueue = new LinkedList<>();
+
+    /**
+     * Map of internal queue id -> deployment instance.
+     */
+    private final Map<String, Deployment> deploymentMap = new HashMap<>();
+
+    /**
+     * <p>If the offered deployment id is unique, then insert the offered deployment at the tail of the queue.</p>
+     *
+     * <p>When an offered deployment has the same deployment id as an already-enqueued element:
+     *     - if the offered deployment meets replacement criteria, then the enqueued element is replaced by the offered
+     *       element, preserving queue order.
+     *     - otherwise, the offered deployment is ignored.</p>
      *
-     * @param deployment deployment
+     * <p>Replacement criteria are as follows:
+     *     - offered.getDeploymentStage().getPriority() > enqueued.getDeploymentStage().getPriority(), OR
+     *     - offered.isCancelled() == true, OR
+     *     - offered.getDeploymentType().equals(SHADOW)</p>
+     *
+     * <p>When an offered deployment is a shadow deployment:
+     *     - if there is already a shadow deployment in the queue, then replace it and preserve queue order.
+     *     - otherwise, enqueue normally.</p>
+     *
+     * @param offeredDeployment the offered deployment instance.
+     * @return true if the queue was modified, otherwise false.
      */
-    public synchronized boolean offer(Deployment deployment) {
-        //For shadow deployment when desired state is reverted, it can result in scheduling a deployment which is
-        // same as
-        if (!DeploymentType.SHADOW.equals(deployment.getDeploymentType()) && deploymentsQueue.contains(deployment)) {
+    public synchronized boolean offer(Deployment offeredDeployment) {
+        if (offeredDeployment == null) {
+            return false;
+        }
```

Comment:
nit: Queue's generally throw NPE if you offer null, or if they allow null as a value they'll process it...any reason to not throw NPE here?

Seems like we should check the deployment is not null before doing anything in the queue, rather than silently allowing usage of it (even if we report the queue is not modified)

Reply:
I was actually inspired by the Queue javadoc here (https://docs.oracle.com/javase/7/docs/api/java/util/Queue.html), which states that, by convention, `offer(..)` methods return a special value on failure, whereas `add(..)` methods throw an exception.

Reply:
https://docs.oracle.com/javase/7/docs/api/java/util/Queue.html#offer(E) throws NPE

the special value is intended to be used when offer would violate constraints (like if you had a fixed size)

I'd say that offering null is not a 'normal' failure case

Reply:
Touche, I now agree